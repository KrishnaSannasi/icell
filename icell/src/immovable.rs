use std::mem::MaybeUninit;

pub type Owner<Im = MaybeUninit<u8>> = crate::Owner<Immovable<Im>>;
pub type ICell<T> = crate::ICell<Address, T>;

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Immovable<T = MaybeUninit<u8>>(pub T, ());

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Address(usize);

impl Immovable {
    pub const fn new() -> Self {
        Self::with_value(MaybeUninit::uninit())
    }

    pub fn owner() -> Owner {
        Owner::new(Self::new())
    }
}

impl<Im> Immovable<Im> {
    #[allow(clippy::unnecessary_operation)]
    pub const fn with_value(value: Im) -> Self {
        // assert that the size is non-zero
        // any value with a non-zero size will have a unique address
        [()][(std::mem::size_of::<Im>() == 0) as usize];
        unsafe { Self::with_value_unchecked(value) }
    }

    /// # Safety
    ///
    /// the given value must have a unique address (this may not be the case for
    /// zero-sized types)
    pub const unsafe fn with_value_unchecked(value: Im) -> Self {
        Self(value, ())
    }

    pub fn owner_with_value(value: Im) -> Owner<Im> {
        Owner::new(Self::with_value(value))
    }

    pub fn cell<T>(&self, value: T) -> ICell<T> {
        use crate::Identifier;
        self.id().cell(value)
    }
}

impl Address {
    pub fn cell<T>(self, value: T) -> ICell<T> {
        unsafe { ICell::from_raw_parts(self, value) }
    }
}

unsafe impl<Im> crate::Identifier for Immovable<Im> {
    type Id = Address;

    fn id(&self) -> Self::Id {
        Address(self as *const Immovable<Im> as usize)
    }

    fn check_id(&self, &id: &Self::Id) -> bool {
        self.id() == id
    }
}
