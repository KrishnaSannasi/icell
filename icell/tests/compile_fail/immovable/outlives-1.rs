use icell::immovable::Immovable;

fn main() {
    let owner = Immovable::owner();
    let value = owner.cell(10);

    let x = owner.read(&value);

    drop(owner);

    assert_eq!(*x, 10);
}
