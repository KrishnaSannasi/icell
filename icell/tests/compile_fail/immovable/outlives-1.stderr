error[E0505]: cannot move out of `owner` because it is borrowed
  --> $DIR/outlives-1.rs:9:10
   |
7  |     let x = owner.read(&value);
   |             ----- borrow of `owner` occurs here
8  | 
9  |     drop(owner);
   |          ^^^^^ move out of `owner` occurs here
10 | 
11 |     assert_eq!(*x, 10);
   |     ------------------- borrow later used here
